### PR TITLE
[core] Graceful handling of returning bundles when node is removed

### LINKED ISSUE #34508: [air] `tune_worker_fault_tolerance_test` is failing
https://buildkite.com/ray-project/release-tests-branch/builds/1568#01879142-b943-4c90-bae3-4d97254a660d

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

When a node is dead, we did a series of clean up for the node. There was [failure like this](https://github.com/ray-project/ray/issues/34508#issuecomment-1520697663) where a node was already removed from the internal data structures (i.e. `ClusterResourceManager.nodes_`) when trying to clean up bundles from the node. 

E.g.:
```
[2023-04-21 15:59:04,500 E 223 235] (gcs_server) ray_syncer-inl.h:305: Failed to read the message from: 3fd2c4710521be4a5d038c6f535bdbf7780e6b39e2707ce56ba2431f
[2023-04-21 16:00:11,585 C 223 223] (gcs_server) container_util.h:88: Key 4825743636673892829 doesn't exist
*** StackTrace Information ***
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x9d9bea) [0x55879aa24bea] ray::operator<<()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x9db6c2) [0x55879aa266c2] ray::SpdLogMessage::Flush()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x9db9d7) [0x55879aa269d7] ray::RayLog::~RayLog()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x395cd7) [0x55879a3e0cd7] ray::map_find_or_die<>()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x395d15) [0x55879a3e0d15] ray::ClusterResourceManager::GetNodeResources()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x30dfbe) [0x55879a358fbe] ray::gcs::GcsPlacementGroupScheduler::TryReleasingBundleResources()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x30eb6a) [0x55879a359b6a] ray::gcs::GcsPlacementGroupScheduler::ReturnBundleResources()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x310158) [0x55879a35b158] ray::gcs::GcsPlacementGroupScheduler::OnAllBundlePrepareRequestReturned()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x31225c) [0x55879a35d25c] std::_Function_handler<>::_M_invoke()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x37e9bb) [0x55879a3c99bb] ray::rpc::ClientCallImpl<>::OnReplyReceived()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x207f95) [0x55879a252f95] std::_Function_handler<>::_M_invoke()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x5abe66) [0x55879a5f6e66] EventTracker::RecordExecution()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x55791e) [0x55879a5a291e] std::_Function_handler<>::_M_invoke()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0x557d96) [0x55879a5a2d96] boost::asio::detail::completion_handler<>::do_complete()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0xacc9ab) [0x55879ab179ab] boost::asio::detail::scheduler::do_run_one()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0xacea99) [0x55879ab19a99] boost::asio::detail::scheduler::run()
/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/gcs/gcs_server(+0xacef52) [0x55879ab19f52] boost::asio::io_context::run()

```

I believe the inconsistency happens when:
1. gcs resouce manager removes the node from the cluster resource manager: https://github.com/ray-project/ray/blob/7ee90897f0ca5db9a6dc009866a7ad3a41736fae/src/ray/gcs/gcs_server/gcs_resource_manager.cc#L292
2. the placement group manager tries to release bundles on the node, and querying the cluster resource manager for the node resource: https://github.com/ray-project/ray/blob/8e3e741d9b94a2694e789d4bcdff16e4c27a501c/src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc#L686-L698

Solution: 
- We will check if a node exists in the cluster resource manager - and if not so, we will declaring bundles returned successfully because that would be done at the node removal callback at cluster resource manager when it does so. The placement group schedule doesn't need to do anything. 
- In fact - we are already doing no-op in some of the functions there, e.g.: 
https://github.com/rickyyx/ray/blob/9a400562a2359479d25d5fb62d88eec3eb1ddc29/src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc#L741-L746
 

## Related issue number
Closes https://github.com/ray-project/ray/issues/34508

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rickyyx** [comment]
Open to alternative solutions  - and would love to get consensus on the root cause as well.

**Comment by rkooo567** [comment]
Hmm does it actually fix the issue?

Seems like the root cause is the ReleaseBundle is asynchronous. That said, if this happens

1. Send CommitBundle or ReleaseBundle RPC
2. node removed 
3. the RPC replied and call GetNodeResources is called.

I think we should handle `GetNodeResources` more gracefully instead? What about we just use this API

```
  /// Return resources associated to the given node_id in ret_resources.
  /// If node_id not found, return false; otherwise return true.
  bool GetNodeResources(scheduling::NodeID node_id, NodeResources *ret_resources) const;
```

and handle the case where node is not alive more gracefully? It seems like this API is used at only 3 places in pg scheduler.cc or UpdateNode, which already handles the dead node case.

**Comment by rickyyx** [comment]
> Seems like the root cause is the ReleaseBundle is asynchronous. That said, if this happens

I thought all the RPCs would be posted to a single event loop - so the release bundle calls and remove node will be serialized?

**Comment by rkooo567** [comment]
It is posted to the event loop "after" RPC is replied or received. We cannot control when the RPC is received. That said, I think it is still possible node failure related information comes first and then releaseBundle is replied later? If. this is confusing, we can also talk in person to resolve (it could be that I also may misunderstand something)

**Comment by rickyyx** [comment]
> It is posted to the event loop "after" RPC is replied or received. We cannot control when the RPC is received. That said, I think it is still possible node failure related information comes first and then releaseBundle is replied later? If. this is confusing, we can also talk in person to resolve (it could be that I also may misunderstand something)

Sync offline - we believe it's safe to do no-ops when releasing the bundles if a node is already removed (cluster resource manager should be responsible for removing the node resources).

**Comment by scv119** [comment]
@rkooo567 should approve here.

is there any test cover this case?

File: src/ray/gcs/gcs_server/test/gcs_placement_group_scheduler_test.cc [hunk 1]
```diff
@@ -1405,6 +1410,46 @@ TEST_F(GcsPlacementGroupSchedulerTest, TestWaitingRemovedBundles) {
...
+  ASSERT_TRUE(raylet_clients_[0]->GrantCommitBundleResources());
+  WaitPlacementGroupPendingDone(0, GcsPlacementGroupStatus::FAILURE);
+  WaitPlacementGroupPendingDone(1, GcsPlacementGroupStatus::SUCCESS);
+
+  // Remove the node.
+  RemoveNode(node);
+
+  // Remove the placement group.
+  const auto &placement_group_id = placement_group->GetPlacementGroupID();
+  scheduler_->DestroyPlacementGroupBundleResourcesIfExists(placement_group_id);
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/test/gcs_placement_group_scheduler_test.cc:1446]: This will call `ReturnBundles` and test the codepath changed in the PR.

**Review by rkooo567 — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Failures seem unrelated (they are reverted in the master)

**Comment by rkooo567** [comment]
and nice unit test!

**Pull request merged**